[WIP] #15990 - HTTP client auto redirects

RomanIakovlev opened the pull request:
This PR adds support for Akka HTTP client to automatically follow the redirects, if it's allowed by RFC. In its current form, it supports redirects to the same host:port, but not the cross-origin redirects, because it re-uses the same underlying `OutgoingConnectionBlueprint` instance as the one being requested by the HTTP client API user originally.

It's not yet fully tested because of the aforementioned limitation (in other words, I don't know if it makes sense to proceed with the current approach), however some tests are available in the `akka.http.impl.engine.client.RedirectTestSpec`. Most basic cases are covered. Particularly, the last one (marked as ignored) highlights the problem with cross-origin redirects.

The settings API (scaladsl/javadsl) and logic related to change of HTTP methods based on the redirect's status code might need some more work.

@ktoso please review.

RomanIakovlev committed:
WIP on #15990: auto redirect for HTTP client

A very naive first attempt of implementing the custom BidiShape for redirection handling,
that sits on top of the Http.ClientLayer. At least
akka.http.impl.engine.client.HighLevelOutgoingConnectionSpec is failing because of it.

RomanIakovlev committed:
WIP on #15990: auto redirect for HTTP client

Current implementation now passes 2 of 4 tests from
akka.http.impl.engine.client.HighLevelOutgoingConnectionSpec.

RomanIakovlev committed:
WIP on #15990: auto redirect for HTTP client

A bit more ceremony around backpressure and availability, although
redirect test is still not working.

RomanIakovlev committed:
WIP on #15990: auto redirect for HTTP client

Totally new implementation of RedirectSupportStage. It now actually
seem to handle redirects well, although "catch response stream truncation"
from HighLevelOutgoingConnectionSpec still fails intermittently. Pipelining
is also seems to work.

RomanIakovlev committed:
WIP on #15990: auto redirect for HTTP client

HighLevelOutgoingConnectionSpec now seem to be always green, because of handling of
upstream failures and completions

RomanIakovlev committed:
WIP on #15990: auto redirect for HTTP client

Added support for infinite redirect loop detection and related test

RomanIakovlev committed:
WIP on #15990: auto redirect for HTTP client

Fixed the rest of the HTTP client tests by replacing manual handling of
unprocessed requests and responses with `emit` function, and adding
forgotten cancellation of `requestIn` on `requestOut` completion

RomanIakovlev committed:
WIP on #15990: auto redirect for HTTP client

Added settings for HTTP client automatic redirect following

RomanIakovlev committed:
WIP on #15990: auto redirect for HTTP client

Apply logic from http://tools.ietf.org/html/rfc2616#section-10.3 to select the
HTTP method and decide whether to proceed with auto redirect or not, based on
the HTTP status code of the response.

RomanIakovlev committed:
WIP on #15990: auto redirect for HTTP client

More tests and refactory. One problem found is that redirects to different host:port
cannot be supported with the current approach.

RomanIakovlev committed:
WIP on #15990: auto redirect for HTTP client

Code cleanup before pull request

typesafehub-validator commented:
Hi @RomanIakovlev,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Typesafe Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

ktoso commented:
OK TO TEST

ktoso commented on akka-http-core/src/main/resources/reference.conf:
```diff
@@ -282,14 +307,14 @@ akka.http {
-    max-header-value-length    = 8k
-    max-header-count           = 64
-    max-chunk-ext-length       = 256
-    max-chunk-size             = 1m
```
Please don't :)

RomanIakovlev replied:
Oops, using auto-format on this file was not the best idea... :) Fixed.

RomanIakovlev committed:
WIP on #15990: restored original format in conf file

ktoso commented on akka-http-core/src/main/resources/reference.conf:
```diff
@@ -170,6 +170,31 @@ akka.http {
+
+    # Automatic redirect settings
+    redirect {
+      # `same-origin` and `cross-origin` contain the same set of settings for
```
excellent, thanks - separating config like this looks very good.

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala:
```diff
@@ -0,0 +1,291 @@
+
+  trait OpInPrinter[T <: Inlet[_]] {
+    def print(): Unit
+  }
```
I'll see later if we really need those but in general let's not add ad-hoc in-middle-of-other-class type classes for printing :)

ktoso replied:
(Can stay here while we're discussing the PR still, no worries)

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala:
```diff
@@ -0,0 +1,291 @@
+  /**
+    * in2 is the responses to original messages. If it's a redirect, its not forwarded upstream, but goes downstream
+    */
+  val responseIn: Inlet[HttpResponse] = Inlet("Input2")
```
Please use this pattern for port names: `RedirectSupportStage.responseIn` / `RedirectSupportStage.requestOut` etc

ktoso replied:
(If something fails, the name is only printed - so it's more helpful to include names like this, we're trying to standardise on this pattern now)

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala:
```diff
@@ -0,0 +1,291 @@
+        ev.print()
+      }
+
+      var inFlight = Vector.empty[HttpRequest]
```
It doesn't seem like we're limiting the max nr of in flight messages; would that be a good idea to add here?

RomanIakovlev replied:
This is a good question. I'm still not sure if this collection `inFlight` is needed at all with current approach. I've already raised this question [before](https://github.com/akka/akka/issues/15990#issuecomment-198074583). I think, current implementation of the push-pull logic prevents more than 1 request in flight, i.e. I don't pull from `requestIn` before I send the reply to the previous request to the `responseOut`. It is also consistent with the in flight numbers I see during tests.

So with this current approach (always one request in flight), the `inFlight` collection is not needed. I've implemented it like this because I was a bit confused on this topic, but now it starts to become clearer. I think I should either just have 1 item in flight, and keep it in an `Option` instead of `Vector`, or change push-pull logic and truly support multiple, if limited, number of requests in flight. Which approach is preferable?

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala:
```diff
@@ -0,0 +1,291 @@
+            printy(loopDetector.toString())
+            val loops = loopDetector(inFlight.head)
+            if (loops.contains(redirect.uri)) {
+              throw InfiniteRedirectLoopException
```
I forgot which is better - throw or fail the stage? 
The answer depends on if we want to log this failure... I guess we do.
// cc @drewhk

ktoso replied:
hm, also - would it be useful to include the url in the exception message?

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala:
```diff
@@ -0,0 +1,291 @@
+import scala.util.control.NoStackTrace
+
+object RedirectSupportStage {
+  type RedirectMapper = (ClientAutoRedirectSettings, HttpRequest, HttpResponse) => Option[HttpRequest]
```
Let's make it a normal class instead perhaps?
Note that you could: 

```
case class RedirectMapper(a: ClientAutoRedirectSettings, b: HttpRequest, c: HttpResponse) 
extends (ClientAutoRedirectSettings, HttpRequest, HttpResponse) => Option[HttpRequest] {
 def apply...
```

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala:
```diff
@@ -0,0 +1,291 @@
+          val response = grab(responseIn)
+          val maybeRedirect = redirectMapper(settings, inFlight.head, response)
+          if (maybeRedirect.isDefined) {
+            val redirect = maybeRedirect.get
```
instead of `isDefined + get` please use `maybeRedirect match { case Some(...`

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala:
```diff
@@ -0,0 +1,291 @@
+      })
+
+    }
+}
```
note to self: re-read again

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/settings/ClientAutoRedirectSettingsImpl.scala:
```diff
@@ -0,0 +1,25 @@
+      ClientAutoRedirectSettingsItemImpl.fromSubConfig(root, sameOriginConfig),
+      ClientAutoRedirectSettingsItemImpl.fromSubConfig(root, crossOriginConfig))
+  }
+}
```
great, thanks

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/settings/ClientAutoRedirectSettingsImpl.scala:
```diff
@@ -0,0 +1,25 @@
+  sameOrigin: ClientAutoRedirectSettingsItem,
+  crossOrigin: ClientAutoRedirectSettingsItem) extends akka.http.scaladsl.settings.ClientAutoRedirectSettings
+
+object ClientAutoRedirectSettingsImpl extends SettingsCompanion[ClientAutoRedirectSettingsImpl]("akka.http.client.redirect") {
```
Hm, companion with Impl in name? Sure

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/settings/ClientConnectionSettingsImpl.scala:
```diff
@@ -24,7 +24,8 @@ private[akka] final case class ClientConnectionSettingsImpl(
   socketOptions: immutable.Seq[SocketOption],
-  parserSettings: ParserSettings)
+  parserSettings: ParserSettings,
+  redirectSettings: akka.http.scaladsl.settings.ClientAutoRedirectSettings)
```
AFAICS keeping binary compatibility will force us to write a constructor the same as the previous one with a default value for the new parameter here.

ktoso replied:
(you can check that by running `mimaReportBinaryIssues`

RomanIakovlev replied:
Speaking of binary compatibility, I'm currently struggling with another class, namely `akka.http.impl.settings.ClientConnectionSettingsImpl`. I've been able to create a constructor without the newly added parameter, however `mimaReportBinaryIssues` also complains about other case classes methods like `copy` and `apply`. AFAIK those are compiler generated and I can't override them. Or can I? And should I?

ktoso replied:
The settings are very smartly structured to allow us to use case classes while retaining public API binary compatibility, you'll notice that:

```scala
/** INTERNAL API */
private[akka] final case class ClientConnectionSettingsImpl(
```

So noone is supposed to touch this class => we can change it and filter out MiMa warnings that it has about it.

RomanIakovlev replied:
Thanks, it's good to know! Then how to mark these issues as ignored? 

Besides those I only have 2 errors in binary compatibility:
```
[error]  * abstract method getRedirectSettings()akka.http.javadsl.settings.ClientAut
oRedirectSettings in class akka.http.javadsl.settings.ClientConnectionSettings is pr
esent only in current version
```
and
```
[error]  * abstract method redirectSettings()akka.http.scaladsl.settings.ClientAutoR
edirectSettings in class akka.http.scaladsl.settings.ClientConnectionSettings is pre
sent only in current version
```
These are the newly added methods. Shouldn't they also be ignored for binary compatibility purposes? I mean, adding a method shouldn't impact the binary compatibility, should it?

ktoso replied:
> I mean, adding a method shouldn't impact the binary compatibility, should it?

It depends. Imagine you've added an abstract method (like we do here), to a class which users are meant to extend. Suddenly you broke their code when they try to upgrade.

In this case the classes are lined up such that we can indeed silence this error – as only we implement this interface.

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/util/JavaMapping.scala:
```diff
@@ -172,6 +172,7 @@ private[http] object JavaMapping {
   implicit object ClientConnectionSettings extends Inherited[js.ClientConnectionSettings, akka.http.scaladsl.settings.ClientConnectionSettings]
   implicit object ConnectionPoolSettings extends Inherited[js.ConnectionPoolSettings, akka.http.scaladsl.settings.ConnectionPoolSettings]
   implicit object ParserSettings extends Inherited[js.ParserSettings, akka.http.scaladsl.settings.ParserSettings]
+  implicit object ClientAutoRedirectSettings extends Inherited[js.ClientAutoRedirectSettings, akka.http.scaladsl.settings.ClientAutoRedirectSettings]
```
:+1:

ktoso commented on akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala:
```diff
@@ -258,7 +258,8 @@ class HttpExt(private val config: Config)(implicit val system: ActorSystem) exte
                   settings: ClientConnectionSettings,
                   log: LoggingAdapter = system.log): ClientLayer =
-    OutgoingConnectionBlueprint(hostHeader, settings, log)
+    RedirectSupportStage(settings.redirectSettings).atop(OutgoingConnectionBlueprint(hostHeader, settings, log))
```
I think we could put that into the blueprint directly

2beaucoup replied:
I can think of some more aspects that'll come up (like caching or proxying). Should they all be baked into the main blueprint?

RomanIakovlev replied:
I thought about those other aspects as well. Maybe it makes sense to introduce something like `HighLevelHttpClient`, to wrap all those aspects and conditionally enable or disable them separately based on the settings? Of course, this potential `HighLevelHttpClient` is bigger than this PR, and there's already an issue for it: https://github.com/akka/akka/issues/16856. I just think this particular place, i.e. `atop(OutgoingConnectionBlueprint)`, is the most appropriate place to add such high-level client.

Every such feature incur some overhead, and client doesn't have to pay the price for something he/she doesn't need, like these redirects, for example. In this particular case, I think I can check the settings on both `cross-origin` and `same-origin` branches, and if both are disabled, don't add this `RedirectSupportStage` on top of the `OutgoingConnectionBlueprint` at all.

ktoso commented on akka-http-core/src/test/scala/akka/http/impl/engine/client/RedirectTestSpec.scala:
```diff
@@ -0,0 +1,127 @@
+import scala.concurrent.duration._
+
+class RedirectTestSpec extends AkkaSpec {
+  implicit val materializer = ActorMaterializer(ActorMaterializerSettings(system).withFuzzing(true))
```
fuzzing is enabled by default (see reference.conf in test scope), you can simply `ActorMaterializer()`

ktoso commented on akka-http-core/src/test/scala/akka/http/impl/engine/client/RedirectTestSpec.scala:
```diff
@@ -0,0 +1,127 @@
+        HttpResponse(status = StatusCodes.MovedPermanently, headers = List(Location(s"/r${(c % cycleLength) + 1}")))
+      }, serverHostName, serverPort)
+
+      val x = Source.fromIterator(() ⇒ Iterator.from(1))
```
hm, wouldn't `single()` be enough? The redirects are happening internally, right?

ktoso commented on akka-http-core/src/test/scala/akka/http/impl/engine/client/RedirectTestSpec.scala:
```diff
@@ -0,0 +1,127 @@
+
+      a[RedirectSupportStage.InfiniteRedirectLoopException.type] should be thrownBy Await.result(x, 3.second)
+
+      binding.futureValue.unbind()
```
`futureValue` on the unbind as well :)

RomanIakovlev replied:
Do you mean `binding.futureValue.unbind().futureValue`? I don't see this in previously existing specs using the same approach.

ktoso replied:
Then those other places are likely wrong :)

ktoso commented on akka-http-core/src/test/scala/akka/http/impl/engine/client/RedirectTestSpec.scala:
```diff
@@ -0,0 +1,127 @@
+      binding.futureValue.unbind()
+    }
+
+    "be able to forward selected headers to redirected requests in different origin" /*in Utils.assertAllStagesStopped */ ignore {
```
I'll check out the branch and investigate this one then; likely not this week though – easter happened.
Sorry for the delay!

RomanIakovlev commented:
Thanks a bunch for the review! I'll have a break for a couple of days now, and will address your findings afterwards.

ktoso commented:
Excellent, thank you! Your work looks very through so far, looking forward to figuring out the missing bits and pushing it over the finish line :) Have a nice easter, cheers :rabbit:

2beaucoup commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala:
```diff
@@ -0,0 +1,291 @@
+      }
+    }
+
+  def sameOrigin(u1: Uri, u2: Uri): Boolean = {
```
```scala
def sameOrigin(u1: Uri, u2: Uri) =
  u2.isRelative || u1.isAbsolute && u1.authority == u2.authority
```

2beaucoup commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala:
```diff
@@ -0,0 +1,291 @@
+  val defaultRedirectMapper: RedirectMapper =
+    (s, req, resp) => {
+      if (resp.status.isRedirection) {
+        inferMethod(req, resp).flatMap { method =>
```
The nested `flatMap`s could be replaced by a for-loop (untested):

```scala
def isAllowed(location: Uri) = {
  val setting =
    if (resp.uri.isRelative || req.uri.isAbsolute && req.uri.authority == resp.uri.authority) s.sameOrigin
    else s.crossOrigin
  setting.getAllow
}

for {
  method <- inferMethod(req, resp)
  location <- resp.header[Location]
  if isAllowed(location.uri)
} yield {
  val headers = item.getForwardHeaders match {
    case All => req.headers
    case Zero => Nil
    case Only(hs) => req.headers.filter(h => hs.contains(h.name))
  }
  req.copy(uri = location.uri, headers = headers, method = method)
}
```

RomanIakovlev committed:
Apply code review comments

- Change defaultRedirectMapper to be a class instead of function instance
- Different headers filtering approach
- Rename ports in the graph stage
- Fix premature graph stage completion (while some elements are in flight)
- Add redirects loop value into InfiniteRedirectLoopException
- Update spec for infinite redirect to use a single HttpRequest
- Change code formatting for shorter lines and readability

RomanIakovlev committed:
Simplify auto redirect settings classes

Remove javadsl- and scaladsl-specific ClientAutoRedirectSettingsItem, since there is
nothing Java or Scala specific in it itself. Use ex-ClientAutoRedirectSettingsItemImpl
directly instead in both javadsl and scaladsl ClientAutoRedirectSettings, with added
Java-style methods.

RomanIakovlev committed:
Add MiMa filters

Mark binary compatibility issues as ignored, after discussion with @ktoso

RomanIakovlev committed:
Apply code review comments

Simplify `DefaultRedirectMapper` class' methods

RomanIakovlev committed:
Code review comments applied

Ensure server is unbound by the end of specs in RedirectTestSpec

ktoso commented:
Was not able to get to this yet, slipping to 2.4.4 (which will be released whenever we have javadsl ready and this likely too)

ktoso commented:
Refs #15990

2beaucoup commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala:
```diff
@@ -0,0 +1,299 @@
+
+    def apply(s: ClientAutoRedirectSettings, req: HttpRequest, resp: HttpResponse): Option[HttpRequest] = {
+      for {
+        location <- resp.header[Location] if resp.status.isRedirection
```
I'd keep the cheap `if` check outside the for-loop.

RomanIakovlev replied:
@2beaucoup In case the `if` is outside the `for` comprehension, I'd have to step out from the `Option` monad and provide the `else` block. I don't think it looks better. Here, have a look:

      if (resp.status.isRedirection) {
        for {
          location <- resp.header[Location]
          method <- inferMethod(req, resp)
          item = if (sameOrigin(req.uri, location.uri)) s.sameOrigin else s.crossOrigin
          if item.allow
        } yield {
          val headers = item.forwardHeaders match {
            case All => req.headers
            case Zero => Nil
            case Only(hs) => req.headers.filter(h => hs.exists(h.is))
          }
          req.withUri(location.value).withHeaders(headers).withMethod(method)
        }
      }
      else {
        None
      }
    }

Instead of:

      for {
        location <- resp.header[Location] if resp.status.isRedirection
        method <- inferMethod(req, resp)
        item = if (sameOrigin(req.uri, location.uri)) s.sameOrigin else s.crossOrigin
        if item.allow
      } yield {
        val headers = item.forwardHeaders match {
          case All => req.headers
          case Zero => Nil
          case Only(hs) => req.headers.filter(h => hs.exists(h.is))
        }
        req.withUri(location.value).withHeaders(headers).withMethod(method)
      }

I'd rather keep the existing version in this case. But as you wish. Let me know if you still prefer to have the `if` outside of the `for`.

2beaucoup replied:
I agree that the pure monadic style looks better. But the suggested change saves the unnecessary location header lookup which will always return `None` for all non-redirect responses and is more expensive to calculate than `isRedirection `.

If only if-statements could come first in for-loops. ;)

2beaucoup commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala:
```diff
@@ -0,0 +1,299 @@
+      for {
+        location <- resp.header[Location] if resp.status.isRedirection
+        method <- inferMethod(req, resp)
+        item = if (sameOrigin(req.uri, Uri(location.value))) s.sameOrigin else s.crossOrigin
```
`Uri(location.value)` can be replaced with `location.uri`.

2beaucoup commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala:
```diff
@@ -0,0 +1,299 @@
+  def apply(settings: ClientAutoRedirectSettings) =
+    BidiFlow.fromGraph(new RedirectSupportStage(settings, DefaultRedirectMapper()))
+
+  def apply(settings: ClientAutoRedirectSettings, redirectMapper: RedirectMapper) =
```
Use default value for `redirectMapper` and remove the above method?

2beaucoup commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala:
```diff
@@ -0,0 +1,299 @@
+          redirectMapper(settings, inFlight.head, response) match {
+            case Some(redirect) =>
+              printy(loopDetector.toString())
+              val loops = loopDetector(inFlight.head)
```
There should be no loops in here. ;) Rename to `followed`?

2beaucoup commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/RedirectSupportStage.scala:
```diff
@@ -0,0 +1,299 @@
+    }
+  }
+
+  case class InfiniteRedirectLoopException(loop: immutable.Seq[Uri])
```
Just `RedirectLoopException`?

2beaucoup commented on akka-http-core/src/test/scala/akka/http/impl/engine/client/RedirectTestSpec.scala:
```diff
@@ -0,0 +1,126 @@
+      val result = Source.fromIterator(() ⇒ Iterator.from(1))
+        .take(N)
+        .map(id ⇒ HttpRequest(uri = s"/r$id"))
+        .via(Http().outgoingConnection(serverHostName, serverPort))
```
Would be nice to use `RedirectSupportStage` directly ISO `Http().outgoingConnection` in these tests so we can get rid of the bindings.

RomanIakovlev replied:
@2beaucoup Sorry, I don't understand your proposal. Could you please clarify?

2beaucoup replied:
Sure. `outgoingConnection` produces a `Flow[HttpRequest, HttpResponse, Future[OutgoingConnection]]`. If you provide such a flow yourself you don't need to to bind/unbind:
```scala
val flow = RedirectSupportStage() join Flow[HttpRequest].map(r ⇒ {
  val c = r.uri.toString.reverse.takeWhile(Character.isDigit).reverse.toInt
  if (c % 2 == 0) {
    HttpResponse(entity = c.toString)
  } else {
    HttpResponse(status = StatusCodes.MovedPermanently, headers = List(Location(s"/r${c + 1}")))
  }
})
val N = 100
val result = Source.fromIterator(() ⇒ Iterator.from(1))
  .take(N)
  .map(id ⇒ HttpRequest(uri = s"/r$id"))
  .via(flow)
  .mapAsync(4)(_.entity.toStrict(1.second))
  .map { r ⇒ val s = r.data.utf8String; log.debug(s); s.toInt }
  .runFold(0)(_ + _)
```

RomanIakovlev commented:
One thing has just occurred to me. Please correct me if I'm wrong anywhere in my reasoning.

Every `OutgoingConnectionBlueprint` instance (further referred to just as "outgoing connection") has maximum 1 corresponding TCP connection at any time. This means, one outgoing connection can only be connected to 1 host. This, in turn, means it's impossible to implement cross-origin redirects inside single outgoing connection.

Next, it seems that all of the existing Akka HTTP client features are implemented as graphs. To send request to another origin means to instantiate the outgoing connection inside another outgoing connection. Or, in other words, to support cross-origin redirects we must change the graph shape in the runtime. Which is not possible, at least currently, if I understand Akka Streams correctly.

To summarize, I think such high-level HTTP client features as redirects support cannot be implemented purely in Akka Streams terms. I think it's should possible to achieve the desired behavior using pure actors, but I'm not sure it's a good way to go forward. Or, maybe it's possible to come up with some creative solution, like changing the graph shape at runtime somehow. 

I hope core team members could shed some light here.

2beaucoup commented:
You are right @RomanIakovlev and that's one more reason to keep this kind of functionality out of the low level blueprints (and even the default connection methods in `Http`?).

I think your stage should work for cross origin redirects by using it atop `Http.superPool`. Atop `outgoingConnection` and `cachedHostConnectionPool` it should work at least for same origin redirects.

drewhk commented:
I think on the per-request API level it can be implemented on the top using Future composition by simply recovering from the case where a redirect has been found with another request Future using flatMap.

In the case of Flow based APIs, a loop might be necessary, reinjecting redirected requests to the wrapped flow (with an upper limit) using the graph DSL

ktoso commented:
Hey there, thanks for the nice chat on ScalaDays :-)
AFAIR we agreed to close it for now, but I'd hope you'd be able to give it another shot given what we've learned from the initial attempt?

A contrib of this feature would be very awesome!
